crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -1,43 +1,92 @@
 };
-use crate::semantic_index::{symbol_table, SemanticIndex};
-use crate::types::{infer_types, ClassType, FunctionType, Name, Type, UnionTypeBuilder};
 use crate::Db;
+use ruff_db::parsed::parsed_module;
 
-/// The inferred types for a single scope.
+use crate::semantic_index::semantic_index;
+use crate::semantic_index::symbol::NodeWithScopeKind;
+
+/// Infer all types for a [`Definition`] (including sub-expressions).
+/// Use when resolving a symbol name use or public type of a symbol.
+#[salsa::tracked(return_ref)]
+pub(crate) fn infer_definition_types<'db>(
+    db: &'db dyn Db,
+    definition: Definition<'db>,
+) -> TypeInference<'db> {
+    let _span = tracing::trace_span!("infer_definition_types", ?definition).entered();
+
+    let index = semantic_index(db, definition.file(db));
+
+    TypeInferenceBuilder::new(db, InferenceRegion::Definition(definition), index).finish()
+}
+
+/// Infer all types for an [`Expression`] (including sub-expressions).
+/// Use rarely; only for cases where we'd otherwise risk double-inferring an expression (RHS of an
+/// assignment, which might be unpacking/multi-target and thus part of multiple definitions, or a
+/// type narrowing guard expression (e.g. if statement test node).
+#[allow(unused)]
+#[salsa::tracked(return_ref)]
+pub(crate) fn infer_expression_types<'db>(
+    db: &'db dyn Db,
+    expression: Expression<'db>,
+) -> TypeInference<'db> {
+    let _span = tracing::trace_span!("infer_expression_types", ?expression).entered();
+
+    let index = semantic_index(db, expression.file(db));
+
+    TypeInferenceBuilder::new(db, InferenceRegion::Expression(expression), index).finish()
+}
+
+/// Infer all types for a [`ScopeId`], including all definitions and expressions.
+/// Use when checking a scope, or needing to provide a type for an arbitrary expression in the
+/// scope.
+#[salsa::tracked(return_ref)]
+pub(crate) fn infer_scope_types<'db>(db: &'db dyn Db, scope: ScopeId<'db>) -> TypeInference<'db> {
+    let _span = tracing::trace_span!("infer_scope_types", ?scope).entered();
+
+    let file = scope.file(db);
+    // Using the index here is fine because the code below depends on the AST anyway.
+    // The isolation of the query is by the return inferred types.
+    let index = semantic_index(db, file);
+
+    TypeInferenceBuilder::new(db, InferenceRegion::Scope(scope), index).finish()
+}
+
+/// A region within which we can infer types.
+pub(crate) enum InferenceRegion<'db> {
+    Expression(Expression<'db>),
+    Definition(Definition<'db>),
+    Scope(ScopeId<'db>),
+}
+
+/// The inferred types for a single region.
 #[derive(Debug, Eq, PartialEq, Default, Clone)]
 pub(crate) struct TypeInference<'db> {
-    /// The types of every expression in this scope.
-    expressions: IndexVec<ScopedExpressionId, Type<'db>>,
+    /// The types of every expression in this region.
+    expressions: FxHashMap<ScopedExpressionId, Type<'db>>,
```

Comment:
Not for this PR: But I wonder if we could have different backing storages depending on the region. Because the scope-level inference could still use an `IndexVec` (and it probably makes sense for it to do so because there are many expressions, to the point where having becomes expensive)

Reply:
Yes, I thought about this as well. I think it is definitely possible, and we can experiment with it and see how much difference it makes.